zrml/swaps/src/lib.rs

```diff
@@ -480,6 +484,12 @@ mod pallet {
             let params = PoolJoinWithExactAmountParams {
                 asset,
                 asset_amount: |asset_balance: BalanceOf<T>, total_supply: BalanceOf<T>| {
+                    let mul: BalanceOf<T> = bmul(
+                        total_supply.saturated_into(),
+                        T::MaxOutRatio::get().saturated_into(),
+                    )?
+                    .saturated_into();
+                    ensure!(pool_amount <= mul, Error::<T>::MaxOutRatio);
                     let asset_amount: BalanceOf<T> = crate::math::calc_single_in_given_pool_out(
                         asset_balance.saturated_into(),
                         Self::pool_weight_rslt(&pool, &asset)?,
```

Comment:
I think those constraints also have to be implemented for `pool_exit_with_exact_asset_amount` and `pool_join_with_exact_asset_amount`